src/Features/CSharp/Portable/Completion/CompletionProviders/OperatorsAndIndexer/OperatorIndexerCompletionProviderBase.cs

```diff
@@ -0,0 +1,167 @@
+        // \uffdc http://www.fileformat.info/info/unicode/char/ffdc/index.htm (non-surrogate)
+        protected const string SortingPrefix = "\uFFFD";
+
+        internal override ImmutableHashSet<char> TriggerCharacters => ImmutableHashSet.Create('.');
+
+        internal override bool IsInsertionTrigger(SourceText text, int insertedCharacterPosition, OptionSet options)
+            => text[insertedCharacterPosition] == '.';
+
+        protected override Task<CompletionDescription> GetDescriptionWorkerAsync(Document document, CompletionItem item, CancellationToken cancellationToken)
+            => SymbolCompletionItem.GetDescriptionAsync(item, document, cancellationToken);
+
+        protected static ImmutableDictionary<string, string> CreatePropertiesBag(params (string key, string value)[] properties)
+        {
+            var builder = ImmutableDictionary.CreateBuilder<string, string>();
+            foreach (var (key, value) in properties)
+            {
+                builder.Add(key, value);
+            }
+
+            return builder.ToImmutable();
+        }
+
+        public override async Task ProvideCompletionsAsync(CompletionContext context)
+        {
+            var cancellationToken = context.CancellationToken;
+            var document = context.Document;
+            var position = context.Position;
+            var root = await document.GetRequiredSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+            var (_, potentialDotToken) = FindTokensAtPosition(position, root);
+            if (!potentialDotToken.IsKind(SyntaxKind.DotToken))
+            {
+                return;
+            }
+
+            var expression = GetParentExpressionOfToken(potentialDotToken);
+            if (expression is null)
+            {
+                return;
+            }
+
+            var semanticModel = await document.ReuseExistingSpeculativeModelAsync(expression.SpanStart, cancellationToken).ConfigureAwait(false);
+            var container = semanticModel.GetTypeInfo(expression, cancellationToken).Type;
+            if (container is null)
+            {
+                return;
+            }
+
+            var completionItems = GetCompletionItemsForTypeSymbol(container, semanticModel, position);
+            context.AddItems(completionItems);
+        }
+
+        protected static (SyntaxToken tokenAtPosition, SyntaxToken potentialDotTokenLeftOfCursor) FindTokensAtPosition(int position, SyntaxNode root)
+        {
+            var tokenAtPosition = root.FindTokenOnLeftOfPosition(position, includeSkipped: true);
+            var potentialDotTokenLeftOfCursor = tokenAtPosition.GetPreviousTokenIfTouchingWord(position);
+            return (tokenAtPosition, potentialDotTokenLeftOfCursor);
+        }
+
+        protected static SyntaxNodeOrToken? FindNodeOrTokenToRemoveAtCursorPosition(SyntaxToken tokenAtCursor)
```

Comment:
i don't get why this returns `SyntaxNodeOrToken?` instead of `SyntaxToken?`.

Reply:
This can be a node (IdentifierNameSyntax: foo) 
```CS
c.fl$$
// committed. "fl" is removed
((float)c).
```
or a Keyword token
```CS
c.case$$
// committed. "case" is removed
((caseClass)c).
```

Reply:
but for the `fl` case, that's still a token.  So why do we ever need to care about nodes?  jsut return the token that matters.  it may be an identifier-token or a keyword-token.  I don't see the reason we'd need the node-parent in teh case of an identifier-token.

Reply:
Fixed in a9c827a1114737f769fb741a9d445fd47a1ecfe2